### PR TITLE
feat: Adding batch deletion function for routing pages

### PR DESCRIPTION
**Pull request by FangSen9000**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

Please update this section with detailed description.

[批量删除.webm](https://user-images.githubusercontent.com/72308243/180745383-d0b5e450-5d33-44ac-a5a7-c49e5d9b4cb3.webm)
You can watch this video for a preview.

On the routing page, the user cannot delete in the state of select all or batch delete. I added related operations.
image

When selecting multiple or single items, a footer bar will pop up for batch deletion of table data. The deselect button will appear in the title of the table.

It also has full and anti - full pages
image

**Related issues**

#2565
#2562

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [x] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

**Comment by SkyeYoung** [comment]
@FangSen9000 Any update here?

**Comment by FangSen9000** [comment]
> @FangSen9000 Any update here?
I've had a day off, and I'll try to update it tonight

File: web/src/pages/Route/List.tsx [hunk 1]
```diff
@@ -118,6 +111,25 @@ const Page: React.FC = () => {
       );
     });
   };
+
```
Comment (SkyeYoung) [inline web/src/pages/Route/List.tsx:114]:
```suggestion
```

Please remove the extra spaces

File: web/src/pages/Route/List.tsx [hunk 2]
```diff
@@ -118,6 +111,25 @@ const Page: React.FC = () => {
       );
     });
   };
+  
+  const handleRemove = async (selectedRows) => {
+    const hide = message.loading('正在删除');
+    if (!selectedRows) return true;
+
```
Comment (SkyeYoung) [inline web/src/pages/Route/List.tsx:118]:
```suggestion
```

ditto

**Review by SkyeYoung — CHANGES_REQUESTED** [review]

**Comment by SkyeYoung** [comment]
image
e2e test also needs to be fixed 🤔

**Comment by FangSen9000** [comment]
> image e2e test also needs to be fixed 🤔
I really want to see if it works. I just built it in Git Bash and prompted the error code 500. This is the problem of the server. Simply using yarn start is not enough. I'm looking for node JS alternatives

**Comment by SkyeYoung** [comment]
> > image e2e test also needs to be fixed 🤔
> > I really want to see if it works. I just built it in Git Bash and prompted the error code 500. This is the problem of the server. Simply using yarn start is not enough. I'm looking for node JS alternatives

Are you developing with Windows? Why not try Linux, you can use a tool such as VMWare to install. Otherwise some problems may be difficult to reproduce. 🤔

**Comment by FangSen9000** [comment]
> > > image e2e test also needs to be fixed 🤔
> > > I really want to see if it works. I just built it in Git Bash and prompted the error code 500. This is the problem of the server. Simply using yarn start is not enough. I'm looking for node JS alternatives
> 
> Are you developing with Windows? Why not try Linux, you can use a tool such as VMWare to install. Otherwise some problems may be difficult to reproduce. 🤔

I used to use the docker recommended by the document, but I couldn't get the latest screenshot due to many small obstacles (maybe the first version can work). I wanted to use node.js yesterday starts with yarn start, which says there is a problem with the server. I really should try Linux that more people choose. I have virtual machine software, but it's a little cumbersome, so it's useless.

**Comment by SkyeYoung** [comment]
> I used to use the docker recommended by the document, but I couldn't get the latest screenshot due to many small obstacles (maybe the first version can work). I wanted to use node.js yesterday starts with yarn start, which says there is a problem with the server. I really should try Linux that more people choose. I have virtual machine software, but it's a little cumbersome, so it's useless.

It is better to try it, it can solve a lot of possible problems

**Comment by SkyeYoung** [comment]
image 

This image doesn't look like a screenshot of APISIX Dashboard, please replace it.

**Comment by SkyeYoung** [comment]
Any update here?

**Comment by SkyeYoung** [comment]
Did you add the English translation?

File: web/src/pages/Route/List.tsx [hunk 3]
```diff
@@ -542,6 +545,27 @@ const Page: React.FC = () => {
...
+          <Space size={24}>
+            <span>
+            {formatMessage({ id: 'page.route.chosen' })} {selectedRowKeys.length} {formatMessage({ id: 'page.route.item' })}
+            </span>
+          </Space>
+        )}
+        tableAlertOptionRender={() => {
+          return (
+            <Space size={16}>
+              <Button
```
Comment (Baoyuantop) [inline web/src/pages/Route/List.tsx:559]: A small suggestion, for the delete type of button we tend to use red to keep the style consistent, for this delete all the dangerous operations, it is best to add a pop-up window to confirm the second

**Review by Baoyuantop — COMMENTED** [review]
Also need to add i18n files for other languages

File: web/src/pages/Route/List.tsx (hunk 3 above)
Comment (FangSen9000) [reply web/src/pages/Route/List.tsx:559]: image
In fact, in Antd UI, deletion generally prefers White，when I have spare power, I will try to add a secondary message, and then open a small patch PR

Comment (Baoyuantop) [reply web/src/pages/Route/List.tsx:559]: Please create an issue to track this issue

**Comment by Baoyuantop** [comment]
Please create an E2E test to cover this behavior

**Comment by FangSen9000** [comment]
> Please create an E2E test to cover this behavior

OK, I have opened the relevant issues and will complete them later

**Comment by Baoyuantop** [comment]
Generally, we want tests and features in one PR

**Comment by SkyeYoung** [comment]
@FangSen9000 Is this ready to be reviewed?

**Comment by FangSen9000** [comment]
> @FangSen9000 Is this ready to be reviewed?

Always be ready

**Comment by FangSen9000** [comment]
> Generally, we want tests and features in one PR

The current new features will not cause errors in the test. I found that several powerful people have written about the test coverage, but their lint problem can't be submitted for the time being. I suggest that let those professionals open PR to study it.

**Comment by SkyeYoung** [comment]
@FangSen9000 Hi, here's [Markdown Guide](https://www.markdownguide.org/getting-started/). Maybe u need this. 😉

**Review by SkyeYoung — APPROVED** [review]

**Review by SkyeYoung — CHANGES_REQUESTED** [review]
Pls add front end e2e tests

**Comment by FangSen9000** [comment]
> Pls add front end e2e tests

@Baoyuantop 
image

I added the front-end E2E test case and deleted the successful promotion, and the detection is normal

File: web/cypress/integration/route/batch-delete-route.spec.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    deleteRouteSuccess: 'Delete Route Successfully',
+    test: 'test',
+    test0: 'test0',
+    test1: 'test1',
+    test2: 'test2',
+    testx: 'testx',
+    desc0: 'desc0',
+    desc1: 'desc1',
+    desc2: 'desc2',
+    value0: 'value0',
```
Comment (Baoyuantop) [inline web/cypress/integration/route/batch-delete-route.spec.js:65]: Why do we need these? I didn't find out what it does.

File: web/cypress/integration/route/batch-delete-route.spec.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      cy.contains('Submit').click();
+      cy.contains(data.submitSuccess);
+      cy.contains('Goto List').click();
+      cy.url().should('contains', 'routes/list');
+    }
+  });
+
+  it('should delete the route', function () {
+    cy.visit('/routes/list');
+    cy.get('[type="checkbox"]').first().check();
```
Comment (Baoyuantop) [inline web/cypress/integration/route/batch-delete-route.spec.js:116]: This selector is not very reliable. Is there a better option? You can refer to https://docs.cypress.io/api/commands/get

**Comment by FangSen9000** [comment]
@Baoyuantop This time, I used the selector to select the route name created, and even if more selection boxes appear in the future, it will not affect this test case. I deleted some useless selector attributes

File: web/cypress/integration/route/batch-delete-route.spec.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      });
+      cy.get(selector.nodes_0_port).type(data.port);
+      cy.get(selector.nodes_0_weight).type(data.weight);
+      cy.contains('Next').click();
+      cy.contains('Next').click();
+      cy.contains('Submit').click();
+      cy.contains(data.submitSuccess);
+      cy.contains('Goto List').click();
+      cy.url().should('contains', 'routes/list');
+    }
```
Comment (guoqqqi) [inline web/cypress/integration/route/batch-delete-route.spec.js:100]: It is best to assert that 3 corresponding routes were successfully created

File: web/cypress/integration/route/batch-delete-route.spec.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    cy.visit('/routes/list');
+    cy.contains(data.test0).get('[type="checkbox"]').check();
+    cy.contains(data.test2).get('[type="checkbox"]').check();
+    cy.contains('BatchDeletion Routes').should('be.visible').click({ timeout });
+    cy.get(selector.notification).should('contain', data.deleteRouteSuccess);
+    cy.get(selector.notificationClose).should('be.visible').click({
+      force: true,
+      multiple: true,
+    });
+  });
```
Comment (guoqqqi) [inline web/cypress/integration/route/batch-delete-route.spec.js:113]: It is better to add an assertion to prove that all data has been deleted

File: web/src/pages/Route/List.tsx [hunk 8]
```diff
@@ -542,6 +545,33 @@ const Page: React.FC = () => {
...
+               onClick={async () => {
+               await remove(selectedRowKeys).then(() => {
+                handleTableActionSuccessResponse(
+                  `${formatMessage({ id: 'component.global.delete' })} ${formatMessage({
+                    id: 'menu.routes',
+                  })} ${formatMessage({ id: 'component.status.success' })}`,
```
Comment (guoqqqi) [inline web/src/pages/Route/List.tsx:565]: It is better not to use this method for internationalization, Chinese and English are hardly compatible at the same time

File: web/cypress/integration/route/batch-delete-route.spec.js (hunk 6 above)
Comment (FangSen9000) [reply web/cypress/integration/route/batch-delete-route.spec.js:100]: ok

**Comment by FangSen9000** [comment]
@Baoyuantop @guoqqqi Do you have any questions or comments on this test case? If not, you can approve it, and then let the back-end testers debug their use cases

File: web/src/pages/Route/List.tsx (hunk 8 above)
Comment (guoqqqi) [reply web/src/pages/Route/List.tsx:565]: image
What I mean is that we shouldn't use template splicing, it leaves extra spaces in the Chinese state.
```suggestion
                  {formatMessage({ id: 'component.global.delete.routes.success' })},
```

**Comment by guoqqqi** [comment]
image
A good tip is that usually the content of these reviews should not be marked as resolved by you, but by the person who reviewed it for you 😄

**Comment by FangSen9000** [comment]
ok




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Qi ***@***.***&gt;; 
发送时间: 2022年8月7日(星期天) 晚上10:29
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/apisix-dashboard] feat: Adding batch deletion function for routing pages (PR #2502)







一个很好的提示是，通常这些评论的内容不应该被标记为由你解决，而是由为你审核的人😄

—
直接回复此邮件，在GitHub上查看，或取消订阅.
***@***.***与&gt;

**Comment by FangSen9000** [comment]
> image A good tip is that usually the content of these reviews should not be marked as resolved by you, but by the person who reviewed it for you 😄

I made some mistakes after I adopted the code. Do you have any good suggestions. Is there any way to solve it? I'm debugging your update on the local machine

**Comment by guoqqqi** [comment]
> > image A good tip is that usually the content of these reviews should not be marked as resolved by you, but by the person who reviewed it for you 😄
> 
> I made some mistakes after I adopted the code. Do you have any good suggestions. Is there any way to solve it? I'm debugging your update on the local machine

you need update the i18n file for `component.global.delete.routes.success`

**Comment by FangSen9000** [comment]
@guoqqqi  I got it on the local machine,  should pass him in as a variable. There was a slight problem with the syntax, which has now been resolved. Thanks for your suggestion.

**Comment by FangSen9000** [comment]
> you need update the i18n file for `component.global.delete.routes.success`
@guoqqqi You are right. This tag is a non-existent tag. I will add this tag tomorrow.
`component.global.delete.routes.success`
You created it...

**Review by guoqqqi — APPROVED** [review]
LGTM

**Comment by FangSen9000** [comment]
@guoqqqi I have completed the changes you suggested. Check that everything is normal
@Baoyuantop @LiteSun  Are there any good questions and suggestions.

File: web/cypress/integration/route/batch-delete-route.spec.js [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    cy.contains(data.test0).should('not.exist');
+    cy.contains(data.test1).should('not.exist');
+    cy.contains(data.test2).should('not.exist');
+    // no match
+    cy.get(selector.nameSearchInput).clear().type(data.testx);
+    cy.contains('Search').click();
+    cy.contains(data.test0).should('not.exist');
+    cy.contains(data.test1).should('not.exist');
+    cy.contains(data.test2).should('not.exist');
+  });
```
Comment (Baoyuantop) [inline web/cypress/integration/route/batch-delete-route.spec.js:141]: Why do we need to test these search functions? We only need to judge that the data does not exist under https://github.com/apache/apisix-dashboard/pull/2502/files#diff-6918beb43c48e71616c503a8dffc6d9cc572f37a69d1e27f792b71fda2e5416aR117.

**Comment by FangSen9000** [comment]
These two assertions remove the name and content of the route, respectively. I think it has three advantages :


1. It can prevent only deleting the key in the route attribute, but not completely delete the route&nbsp;


2. It IS possible that THE TEST case on PAGE 2 will not BE FOUND IF there ARE more test cases later on.


3. The search function automatically refreshes the page, which can prevent timeout errors due to unrendered, and can make the test case more stable




---Original---
From: ***@***.***&gt;
Date: Mon, Aug 8, 2022 09:54 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [apache/apisix-dashboard] feat: Adding batch deletion functionfor routing pages (PR #2502)




 
@Baoyuantop commented on this pull request.
 
 
In web/cypress/integration/route/batch-delete-route.spec.js:
 &gt; +    cy.contains(data.test1).should('not.exist'); +    cy.contains(data.test0).should('not.exist'); +    cy.contains(data.test2).should('not.exist'); +    // partial match +    cy.get(selector.nameSearchInput).clear().type(data.test2); +    cy.contains('Search').click(); +    cy.contains(data.test0).should('not.exist'); +    cy.contains(data.test1).should('not.exist'); +    cy.contains(data.test2).should('not.exist'); +    // no match +    cy.get(selector.nameSearchInput).clear().type(data.testx); +    cy.contains('Search').click(); +    cy.contains(data.test0).should('not.exist'); +    cy.contains(data.test1).should('not.exist'); +    cy.contains(data.test2).should('not.exist'); +  });  
Why do we need to test these search functions? We only need to judge that the data does not exist under https://github.com/apache/apisix-dashboard/pull/2502/files#diff-6918beb43c48e71616c503a8dffc6d9cc572f37a69d1e27f792b71fda2e5416aR117.

**Review by bzp2010 — APPROVED** [review]

**Review by SkyeYoung — APPROVED** [review]

**Review by Baoyuantop — APPROVED** [review]

**Pull request merged**